### DISCUSSION TITLE
Set custom url path with nginx reverse proxy

### DISCUSSION THREAD

**Discussion by cvc90** [seriousm4x/UpSnap#367] [Feature requests]
Could you tell me how to get UpSnap to work correctly with an nginx reverse proxy? Do I have to add some special parameter to make it work?

location ^~ /UpSnap {		
proxy_pass http://localhost:8090/;		
}

As you can see in the images, it does not do the redirections correctly

![image](https://user-images.githubusercontent.com/76731844/216962022-d4da393f-7417-4cc2-819a-a44ad74d4940.png)
![image](https://user-images.githubusercontent.com/76731844/216962106-69827b35-bc7e-4bbf-b321-befb4a33f236.png)

**Comment by seriousm4x** [comment]
I think you are trying to use a sub route. Thats not supported at the moment. Please use a subdomain instead

**Comment by cvc90** [comment]
For security/organizational reasons I can't use a subdomain, I need to use a secondary path.

Would it be possible to include this option for future versions?

**Comment by seriousm4x** [comment]
Currently not. It requires the frontend to adjust for the custom path (not that much of a deal), but also the backend, which is not really supported yet https://github.com/pocketbase/pocketbase/discussions/1802#discussioncomment-4892902

**Comment by cvc90** [comment]
Ok, well, we hope that one day that function will arrive.

> Currently not. It requires the frontend to adjust for the custom path (not that much of a deal), but also the backend, which is not really supported yet [pocketbase/pocketbase#1802 (reply in thread)](https://github.com/pocketbase/pocketbase/discussions/1802#discussioncomment-4892902)

Another alternative that could be valid for me would be to configure UpSnap to load https natively, would it be possible to configure UpSnap to load https natively?

**Comment by seriousm4x** [comment]
You can already use https with automatic Lets Encrypt certs:
```
> upsnap.exe serve --help
Starts the web server (default to 127.0.0.1:8090)

Usage:
  upsnap.exe serve [flags]

Flags:
  -h, --help              help for serve
      --http string       api HTTP server address (default "127.0.0.1:8090")
      --https string      api HTTPS server address (auto TLS via Let's Encrypt)
                          the incoming --http address traffic also will be redirected to this address
      --origins strings   CORS allowed domain origins list (default [*])

Global Flags:
      --debug                  enable debug mode, aka. showing more detailed logs
      --dir string             the PocketBase data directory (default "C:\\Users\\max\\Downloads\\UpSnap_3.0.3_windows_amd64\\pb_data")
      --encryptionEnv string   the env variable whose value of 32 characters will be used
                               as encryption key for the app settings (default none)
```
https://pocketbase.io/docs/going-to-production/#deployment-strategies

**Comment by cvc90** [comment]
Can I use my own certificates? Since the Lets Encrypt certificates would not work for me from within the VPN / LAN

**Comment by seriousm4x** [comment]
Does not look like it
https://github.com/pocketbase/pocketbase/issues/957
https://github.com/pocketbase/pocketbase/discussions/930#discussioncomment-4054165

**Comment by cvc90** [comment]
You don't seem to see a feasible solution to the problem of using certificates are redirected to using a reverse proxy which for security/organizational reasons, I can't use a subdomain and need to use a secondary path.

**Comment by daenney** [comment]
I think it should work if you have a trailing slash on both the `location` and the `proxy_pass` directives if I'm not mistaken? The trailing slash in nginx does a lot of magic.

You may need a `rewrite` in the `location /` to top it off:
```
if ($http_referer ~ "^https?://[^/]+/subdir/") {
    rewrite ^/(.*) https://$http_host/subdir/$1 redirect;
}
```

**Comment by cvc90** [comment]
What you are telling me would be something like this?

```
location ^~ /UpSnap/ {
    proxy_pass http://localhost:8090/;
	if ($http_referer ~ "^https?://[^/]+/subdir/") {
		rewrite ^/(.*) https://$http_host/subdir/$1 redirect;
		}		
}
```

**Comment by daenney** [comment]
Almost. The referrer check needs to go into `location /` to rewrite those requests properly since UpSnap itself won't generate URLs with `/UpSnap` as a prefix. Also you need to repleace the `subdir` in the `if` and `rewrite` like to `UpSnap`

**Comment by cvc90** [comment]
Ok!, What you are telling me would be something like this?

```
location / {		
	root  /volumeX/WWW;
	index index.php index.html index.htm;			
	if ($http_referer ~ "^https?://[^/]+/UpSnap/") {
		rewrite ^/(.*) https://$http_host/UpSnap/$1 redirect;
	}					

location ^~ /UpSnap/ {
    proxy_pass http://localhost:8090/;	
}
```

**Comment by seriousm4x** [comment]
> Ok!, What you are telling me would be something like this?
> 
> ```
> location / {		
> 	root  /volumeX/WWW;
> 	index index.php index.html index.htm;			
> 	if ($http_referer ~ "^https?://[^/]+/UpSnap/") {
> 		rewrite ^/(.*) https://$http_host/UpSnap/$1 redirect;
> 	}					
> 
> location ^~ /UpSnap/ {
>     proxy_pass http://localhost:8090/;	
> }
> ```

Is this a working solution for your issue?

**Comment by johndo100** [comment]
I did it on secondary path and caddy reserve proxy but It didnt work.
Only admin /_/ work.
The client home page didnt show any thing.

I used handle_path in Caddyfile anyway..

**Comment by jordantgh** [comment]
Just want to plus 1 this, it would save a lot of work on the user end making much easier to integrate with personal dashboard etc.

**Comment by seriousm4x** [comment]
I looked at the problem again and found out that this is currently not possible in SvelteKit. You can create a [custom base path](https://kit.svelte.dev/docs/configuration#paths) in SvelteKit, but it has to be specified during the build. 

Unfortunately, this cannot be changed by environment variables as [$env/dynamic/public](https://kit.svelte.dev/docs/modules#$env-dynamic-public) vars need a node runtime (which we don't have in this case) and [$env/static/public](https://kit.svelte.dev/docs/modules#$env-static-public) gets prerendered at build time.

If some Svelte/JS genius has an idea who to solve this, I'm open for solutions but I think it's currently not possible.

**Comment by seriousm4x** [comment]
Turns out, using sub paths is now possible with the `resolve()` function in sveltekit: https://svelte.dev/docs/kit/$app-paths#resolve

This feature is now part of [5.2.0-beta.0](https://github.com/seriousm4x/UpSnap/releases/tag/5.2.0-beta.0). I've added it to the [readme](https://github.com/seriousm4x/UpSnap?tab=readme-ov-file#run-in-sub-path).

**Reply by cvc90** [reply]
Thanks! I'm trying to test this new feature but can't seem to get it right. Would this be correct for an Nginx setup?

location /upsnap-sub-path/ {
proxy_pass http://127.0.0.1:8090;
proxy_set_header Host $host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $scheme;
}

Do I need to specify a path or configuration in Docker?

**Reply by seriousm4x** [reply]
I think the part you are missing is the trailing slash at `proxy_pass`. This config is working for me:

```
http {
    server {
        listen 8091;
        server_name localhost;
        location /upsnap-sub-path/ {
            proxy_pass http://localhost:8090/;
            proxy_redirect off;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
    }
}
```

**Reply by cvc90** [reply]
Thanks for sharing your code!

It seems to load now, but I see a loading error with the gopher.svg image when it displays the login (/UpSnap/login/).

Reviewing the code, I see that one is loaded via a relative URL and the other via an absolute URL.

href="../gopher.svg"/> --> OK
<img src="/gopher.svg" --> FAIL

You would simply need to change <img src="/gopher.svg" to <img src="../gopher.svg" and the error would be fixed.

**Comment by tvup** [comment]
If you're running UpSnap behind a reverse proxy and want WoL to trigger automatically on web traffic (instead of manually), check out [idle-less](https://github.com/tvup/idle-less).

It's a Docker reverse proxy that detects when a backend server is offline and sends WoL magic packets automatically. Visitor hits your domain → WoL → server boots → traffic flows.

UpSnap is great for dashboard/manual control. idle-less handles automated "wake on traffic". Together they cover the full power management lifecycle.

GitHub: https://github.com/tvup/idle-less
Demo: https://tvup.github.io/idle-less/demo.html
